bundles/org.openhab.binding.mynice/src/main/java/org/openhab/binding/mynice/internal/handler/It4WifiHandler.java

```diff
@@ -0,0 +1,243 @@
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.mynice.internal.config.It4WifiConfiguration;
+import org.openhab.binding.mynice.internal.discovery.MyNiceDiscoveryService;
+import org.openhab.binding.mynice.internal.xml.MyNiceXStream;
+import org.openhab.binding.mynice.internal.xml.RequestBuilder;
+import org.openhab.binding.mynice.internal.xml.dto.CommandType;
+import org.openhab.binding.mynice.internal.xml.dto.Device;
+import org.openhab.binding.mynice.internal.xml.dto.Event;
+import org.openhab.binding.mynice.internal.xml.dto.Response;
+import org.openhab.binding.mynice.internal.xml.dto.T4Command;
+import org.openhab.core.config.core.Configuration;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseBridgeHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link It4WifiHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public class It4WifiHandler extends BaseBridgeHandler {
+    private static final int MAX_HANDSHAKE_ATTEMPTS = 3;
+    private static final int KEEPALIVE_DELAY_S = 240; // Timeout seems to be at 6 min
+
+    private final Logger logger = LoggerFactory.getLogger(It4WifiHandler.class);
+    private final List<MyNiceDataListener> dataListeners = new CopyOnWriteArrayList<>();
+
+    private @NonNullByDefault({}) RequestBuilder reqBuilder;
+    private @Nullable It4WifiConnector connector;
+    private final MyNiceXStream xstream = new MyNiceXStream();
+    private List<Device> devices = new ArrayList<>();
+    private @Nullable ScheduledFuture<?> keepAliveJob;
+    private int handshakeAttempts = 0;
+
+    public It4WifiHandler(Bridge thing) {
+        super(thing);
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Set.of(MyNiceDiscoveryService.class);
+    }
+
+    public void registerDataListener(MyNiceDataListener dataListener) {
+        dataListeners.add(dataListener);
+        notifyListeners(devices);
+    }
+
+    public void unregisterDataListener(MyNiceDataListener dataListener) {
+        dataListeners.remove(dataListener);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // we do not handle commands
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        if (getConfigAs(It4WifiConfiguration.class).username.isBlank()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
```

Comment:
Yes because you will not reach this point if they are incorrect. macAddress is mandatory to establish the connection, hostname is needed later in the communication protocol.